### PR TITLE
modules/gnome, modules/Python: Allow injecting RPATH flags through LDFLAGS if needed

### PR DESCRIPTION
**Pull request by amyspark**
Fixes communicating the RPATH to g-i-scanner in macOS.

See #14169

cc @nirbheek

### CODE REVIEW

**Comment by nirbheek** [comment]
Would be great to have a test. You could modify `test cases/frameworks/7 gnome`, perhaps. If you need to inspect the output to check that the LC_RPATH is set, you could add a unit test: `unittests/darwintests.py`.

**Comment by amyspark** [comment]
Added a test for the former (it's a bit noisy, as I needed to add MSVC compatibility; let me know if that's needed or if it can be removed).

For the changes re: RPATH, darwintests.py is not effective for this corner case; the flags need to pass through the GNOME module's filtering. The GNOME-less case is already covered in the `unittests-appleclang` job.

File: mesonbuild/modules/gnome.py [hunk 1]
```diff
@@ -1161,11 +1164,20 @@ def generate_gir(self, state: 'ModuleState', args: T.Tuple[T.List[T.Union[Execut
...
+            self._get_scanner_ldflags(dep_internal_ldflags),
+        ]:
+            scan_internal_ldflags += cli_flags
+            scan_env_ldflags = env_flags
+        for cli_flags, env_flags in [
+            self._get_scanner_ldflags(external_ldflags),
+            self._get_scanner_ldflags(dep_external_ldflags)
+        ]:
+            scan_external_ldflags += cli_flags
+            scan_env_ldflags = env_flags
```
Comment (eli-schwartz) [inline mesonbuild/modules/gnome.py:1180]: This unfortunately looks like it's been deformatted by the "black" code deformatter. Can you fix it?

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: mesonbuild/modules/gnome.py (hunk 1 above)
Comment (amyspark) [reply mesonbuild/modules/gnome.py:1180]: Sure, I'll submit the change with the CI fixes.

Comment (eli-schwartz) [reply mesonbuild/modules/gnome.py]: I'd also probably change the `return` in _get_scanner_ldflags

Note the distinction is not necessarily fitting everything on one line, rather the line breaking algorithm should avoid placing code contained inside parentheses or list notation, as "indented by four lines". Notice in the above diff, that both the values inside `[]` and the indented body of the for block, blend in together.

So, this would be a better way to break across lines:
```python
        for cli_flags, env_flags in (self._get_scanner_ldflags(internal_ldflags),
                                     self._get_scanner_ldflags(dep_internal_ldflags)):
```

because both elements of the tuple are lined up, but both are to the right of the "for x, y in" part.

It also fits in 86 columns, as opposed to 131 columns.

File: mesonbuild/modules/gnome.py [hunk 2]
```diff
@@ -702,14 +702,14 @@ def _get_dependencies_flags_raw(
...
-                            external_ldflags.update([f'-Wl,-rpath {lib_dir}'])
+                            external_ldflags.update([f'-Wl,-rpath,{lib_dir}'])
                         libname = os.path.basename(flag)
                         if libname.startswith("lib"):
                             libname = libname[3:]
                         libname = libname.split(".so")[0]
                         flag = f"-l{libname}"
                     # FIXME: Hack to avoid passing some compiler options in
-                    if flag.startswith("-W"):
+                    if flag.startswith("-W") and not flag.startswith('-Wl,-rpath'):
```
Comment (nirbheek) [inline mesonbuild/modules/gnome.py:712]:
```suggestion
                    if flag.startswith("-W") and not flag.startswith('-Wl,-rpath,'):
```

Otherwise `-Wl,-rpath`  will be allowed through without the RPATH following it.

File: mesonbuild/modules/gnome.py [hunk 3]
```diff
@@ -1089,11 +1091,12 @@ def _get_scanner_cflags(cflags: T.Iterable[str]) -> T.Iterable[str]:
...
     @staticmethod
-    def _get_scanner_ldflags(ldflags: T.Iterable[str]) -> T.Iterable[str]:
+    def _get_scanner_ldflags(ldflags: T.Iterable[str]) -> tuple[list[str], list[str]]:
         'g-ir-scanner only accepts -L/-l; must ignore -F and other linker flags'
-        for f in ldflags:
-            if f.startswith(('-L', '-l', '--extra-library')):
-                yield f
+        return (
+            [f for f in ldflags if f.startswith(('-L', '-l', '--extra-library'))],
+            [f for f in ldflags if f.startswith(('-Wl,-rpath'))],
```
Comment (nirbheek) [inline mesonbuild/modules/gnome.py:1098]:
```suggestion
            [f for f in ldflags if f.startswith(('-Wl,-rpath,'))],
```

Same reason.

**Review by nirbheek — APPROVED** [review]

**Comment by nirbheek** [comment]
PR has conflicts that need fixing.

**Comment by nirbheek** [comment]
LGTM, will merge in a week if there are no objections
